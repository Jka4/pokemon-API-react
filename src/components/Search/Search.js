import React, { lazy, Suspense, useState } from "react";
import { Provider, connect } from "react-redux";
import Loader from "react-loader-spinner";
import { NavLink } from "react-router-dom";
import PropTypes from "prop-types";
import Fuse from "fuse.js";

import store from "@Store";
import { setDelailedPageData } from "@APIutils";

import "./styles/style.scss";

const ImageContainer = lazy(() => import("@ImageContainer"));

let fuseOptions = {
  shouldSort: true,
  tokenize: true,
  matchAllTokens: false,
  findAllMatches: false,
  includeScore: true,
  includeMatches: true,
  threshold: 0.3,
  location: 0,
  distance: 10,
  maxPatternLength: 20,
  minMatchCharLength: 2,
  keys: ["name", "weight", "id"],
};

const Search = (props) => {
  const [showResult, setShowResult] = useState(false);
  const [searchResult, setSearchResult] = useState([]);
  const { pokemonDataArray } = props;

  const handleChange = (event) => {
    const value = event.target.value;

    let fuse = new Fuse(pokemonDataArray, fuseOptions);
    let result = fuse.search(value).slice(0, 11);
    setSearchResult(result);
  };

  const handleFocus = () => {
    setShowResult(true);
  };

  const handleBlur = () => {
    const debounce = () => {
      setShowResult(false);
    };
    setTimeout(debounce, 100);
  };

  let handleClick = (value) => {
    setDelailedPageData(value, true);
  };

  return (
    <div className="search">
      <input
        type="search"
        name="searchInput"
        id="searchInput"
        placeholder="Pickachu, 123, 900g"
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        autoComplete="off"
        aria-label="Search"
      />

      <div className="searchResults">
        {showResult && (
          <ul className="searchList">
            {searchResult.map((i, key) => (
              <NavLink
                key={(i.item.id, key)}
                to={`/detailedPage/pokemon/${i.item.name}`}
                className="searchItem_outer"
              >
                <li
                  data-id={i.item.id}
                  className="searchItem"
                  onClick={() => handleClick(i.item.id)}
                >
                  <span className="item_name">NAME: {i.item.name}</span>
                  <span className="item_id">ID: {i.item.id}</span>
                  <span className="item_weight">WEIGHT: {i.item.weight}</span>
                  <Suspense
                    fallback={
                      <Loader
                        type="TailSpin"
                        height={30}
                        width={30}
                        color={"red"}
                      />
                    }
                  >
                    <ImageContainer
                      url={i.item.image}
                      cn={"searchItem__image"}
                    />
                  </Suspense>
                </li>
              </NavLink>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
};

const ConnectedSearch = connect((store) => {
  return {
    pokemonDataArray: store.pokemonsArr,
  };
})(Search);

export default (props) => (
  <Provider store={store}>
    <ConnectedSearch {...props} />
  </Provider>
);

Search.propTypes = {
  pokemonDataArray: PropTypes.array,
};
