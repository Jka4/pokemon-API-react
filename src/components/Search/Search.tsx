/* eslint-disable import/no-anonymous-default-export */
import React, { lazy, useState } from 'react';
import { Provider, connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import Fuse from 'fuse.js';
import { useDebounce } from 'ahooks';

import store from 'Store/store';

import './styles/style.scss';

const ImageContainer = lazy(() => import('components/ImageContainer/ImageContainer'));

let fuseOptions = {
  shouldSort: true,
  tokenize: true,
  matchAllTokens: false,
  findAllMatches: false,
  includeScore: true,
  includeMatches: true,
  threshold: 0.3,
  location: 0,
  distance: 10,
  maxPatternLength: 20,
  minMatchCharLength: 2,
  keys: ['name', 'weight', 'id'],
};

type SearchProps = {
  pokemonDataArray: Pokes[];
};

type Pokes = {
  id: number;
  name: string;
  weight: number;
  image: string;
  imageHQ: string;
  placeholderBase64: string;
  chain: {
    species_name: string;
  }[];
};

const Search: React.FC<SearchProps> = ({ pokemonDataArray }: SearchProps) => {
  const [showResult, setShowResult] = useState<boolean>(false);
  const [searchResult, setSearchResult] = useState<any[]>([]);
  const debouncedSearchResult = useDebounce(searchResult, { wait: 300 });

  const handleChange = (event: any) => {
    const value: string = event.target.value;

    let fuse = new Fuse(pokemonDataArray, fuseOptions);
    let result = fuse.search(value).slice(0, 11);
    setSearchResult(result);
  };

  const handleFocus = () => {
    setShowResult(true);
  };

  const handleBlur = () => {
    setTimeout(() => {
      setShowResult(false);
    }, 400);
  };

  const fallback = (placeholderBase64: string) => {
    return (
      <>
        <img src={placeholderBase64} className="placeholderBase64 deBlur" alt="placeholderBase64" />
      </>
    );
  };

  return (
    <div className="search">
      <input
        type="search"
        name="searchInput"
        id="searchInput"
        placeholder="Pikachu, 25, 900g"
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        autoComplete="off"
        aria-label="Search"
      />

      <div className="searchResults">
        {showResult && (
          <ul className="searchList">
            {debouncedSearchResult.map((i, key) => (
              <NavLink key={(i.item.id, key)} to={`/detailedPage/pokemon/${i.item.name}`} className="searchItem_outer">
                <li data-id={i.item.id} className="searchItem">
                  <span className="item_name">NAME: {i.item.name}</span>
                  <span className="item_id">ID: {i.item.id}</span>
                  <span className="item_weight">WEIGHT: {i.item.weight}</span>
                  <ImageContainer
                    url={i.item.image}
                    cn={'searchItem__image'}
                    fallback={fallback(i.item.placeholderBase64)}
                  />
                </li>
              </NavLink>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
};

const ConnectedSearch = connect((store: { pokemonsArr: Pokes[] }) => {
  return {
    pokemonDataArray: store.pokemonsArr,
  };
})(Search);

export default (props = {}) => (
  <Provider store={store}>
    <ConnectedSearch {...props} />
  </Provider>
);
