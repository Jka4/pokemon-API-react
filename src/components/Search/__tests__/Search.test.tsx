import Search from '../Search';

import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import POKEMON from 'utils/pokemonDataArray';

const mockStore = configureMockStore([thunk]);

describe('<Search/> ', () => {
  const store = mockStore({ pokemonArr: POKEMON });

  const container = mount(
    <Provider store={store}>
      <Search />
    </Provider>,
  );

  it('renders correctly', () => {
    expect(container).toMatchSnapshot();
  });
});
