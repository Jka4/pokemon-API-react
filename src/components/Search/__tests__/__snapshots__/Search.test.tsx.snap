// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/>  renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Search>
    <styled.div>
      <div
        className="sc-dlfnbm eQOXtf"
      >
        <styled.input
          aria-label="Search"
          autoComplete="off"
          id="searchInput"
          name="searchInput"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Pikachu, 25, 900g"
          type="search"
        >
          <input
            aria-label="Search"
            autoComplete="off"
            className="sc-hKgILt cEFGKr"
            id="searchInput"
            name="searchInput"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Pikachu, 25, 900g"
            type="search"
          />
        </styled.input>
        <styled.div>
          <div
            className="sc-eCssSg jwRzZN"
          />
        </styled.div>
      </div>
    </styled.div>
  </Search>
</Provider>
`;
